### PR TITLE
refactor: use jupyter_server ExtensionApp

### PR DESCRIPTION
**Pull request by maartenbreddels**
A 'reboot' of #270 with a single clean commit to avoid merge conflicts with other PR's.

@Zsailer I took out many code reformatting changes, since they confused the diff-er, and me, and with large outstanding PR's it will lead to many more merge conflicts and a polluted history. I hope you don't hold it against me :)

Replaces #270

### CODE REVIEW

**Comment by Zsailer** [comment]
This is great! Thank you, @maartenbreddels.

**Comment by Zsailer** [comment]
We're rolling back jupyter_server to support Python 3.5 for a new release (0.2) with the ExtensionApp. This should be done by Friday.

Then, we'll likely drop Python 3.5 in a jupyter_server 0.3 release. The new Kernel Provider work from @kevin-bates requires that we're hoping to land in master this next month requires Python 3.6 and above.

**Comment by kevin-bates** [comment]
Just to be clear, we can easily fix the 3.5 support issue in kernel provider stuff.  There was a recent change that unexpectedly introduced a syntax error only on 3.5 - and we had cut the 0.5.0 release just prior to that discovery (thus the hesitancy to fix it for something that we thought was more near to its EOL).  Of course, by the time jupyter-server 0.3.0 is developed, 3.5 may be deprecated, but we shouldn't let this hold up use of JKM.  It's relatively easy to produce a patch release (0.5.1) that restores 3.5 support in JKM if necessary.

**Comment by SylvainCorlay** [comment]
Note: we have a constraint to support Python 3.5 for a little bit more in Voilà. Having jupyter_server be compatible with python 3.5 would be great!

**Comment by Zsailer** [comment]
Ok, we've reverted jupyter_server to support 3.5 in jupyter/jupyter_server#150.

**Comment by maartenbreddels** [comment]
Thanks @kevin-bates , @Zsailer  and @echarles for correcting this before I even had time to reply 👏 

Related:
https://github.com/jupyter/jupyter_server/pull/150
https://github.com/jupyter/jupyter_server/pull/142

**Comment by Zsailer** [comment]
Just release [Jupyter Server 0.2.0](https://github.com/jupyter/jupyter_server/releases).

Still supports Python 3.5 and includes all the ExtensionApp work (with tests, finally). Let us know if you run into any issues. I'm happy to help with this PR too, if you'd like.

File: setup.py [hunk 1]
```diff
@@ -371,12 +371,13 @@ def get_data_files():
...
     'entry_points': {
         'console_scripts': [
+            'jupyter-voila = voila.app:main',
             'voila = voila.app:main'
         ]
     },
     'install_requires': [
         'async_generator',
-        'jupyter_server>=0.1.0,<0.2.0',
+        'jupyter_server>=0.2.0.dev0,<0.3.0',
```
Comment (jtpio) [inline setup.py]: Is `.dev0` still needed now that `0.2.0` has been released?

File: tests/app/conftest.py [hunk 2]
```diff
@@ -25,26 +26,53 @@ def voila_args_extra():
...
+    return [voila_notebook, voila_config_file_paths_arg] + voila_args_extra
 
 
 @pytest.fixture
-def voila_app(voila_args, voila_config):
-    voila_app = VoilaTest.instance()
-    voila_app.initialize(voila_args + ['--no-browser'])
-    voila_config(voila_app)
-    voila_app.start()
+def voila_app(server_config, voila_config, voila_args):
```
Comment (Zsailer) [inline tests/app/conftest.py]: I released `jupyter_server` 0.2.1 (see [changelog](https://github.com/jupyter/jupyter_server/blob/master/CHANGELOG.md)) today, which includes a pytest-plugin for jupyter_server. 

Inside this plugin, we've added a `serverapp` fixture and a `configurable_serverapp` fixture that handle the creation + teardown of a jupyter server instance between each test so you don't have to.

You can see an example using this plugin [here in nbclassic](https://github.com/Zsailer/nbclassic/blob/notebook-handlers/tests/conftest.py).

File: tests/app/execute_test.py [hunk 3]
```diff
@@ -14,8 +14,8 @@
 
 
 @pytest.mark.gen_test
-def test_hello_world(http_client, base_url):
-    response = yield http_client.fetch(base_url)
+def test_hello_world(http_client, default_url):
```
Comment (Zsailer) [inline tests/app/execute_test.py]: The jupyter server pytest-plugin we added in jupyter_server 0.2.1 drastically improves this syntax.

* You can now use `async`/`await` syntax.
* we over a `fetch` fixture—a function for making requests to the jupyter server API and extension handlers.

To see an example, check out this [simple unit test in Jupyter Server](https://github.com/jupyter/jupyter_server/blob/master/tests/services/api/test_api.py).

File: setup.py (hunk 1 above)
Comment (Zsailer) [reply setup.py]: Yes. I'd set the version to `>=0.2.1, <0.3.0`.

**Comment by Zsailer** [comment]
I mentioned in my comments, I released Jupyter Server 0.2.1 today. This includes a pytest-plugin for jupyter server extensions (like future Voila).

This significantly improves the syntax when writing extension tests:
* Can now use `async`/`await` syntax when writing tests (see [example here](https://github.com/jupyter/jupyter_server/blob/master/tests/services/api/test_api.py)).
* `serverapp` fixture that handles setup/teardown. 
* `fetch` fixture for asynchronous/awaitable requests to the serverapp+extension APIs.
* other helpful fixtures, see [summary list here](https://github.com/jupyter/jupyter_server/blob/master/CHANGELOG.md#021---2020-1-10).
* Extend the serverapp fixture with a `voila_app` fixture easily. See [example from nbclassic](https://github.com/Zsailer/nbclassic/blob/notebook-handlers/tests/conftest.py).

To start using today, `pip install jupyter_server>=0.2.1` and add `pytest_jupyter_server` to your top-level `conftest.py`.

**Comment by jtpio** [comment]
This looks great!

Just tried locally with `jupyter_server` 0.2.1. It looks like some minor adjustments are required to take into accounts the latest changes (getting a `KeyError: 'extension_name'` at the moment).

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -47,6 +47,13 @@ jobs:
       env:
         PYTHON_VERSION: ${{ matrix.python-version }}
 
+    - name: Install dev/master versions
+      run: |
+       source "$CONDA/etc/profile.d/conda.sh"
+       conda activate voila-tests
+       pip uninstall jupyter_server -y
+       pip install git+https://github.com/jupyter/jupyter_server@master
```
Comment (jtpio) [inline .github/workflows/main.yml]: We should be able to install the released version now?

File: setup.py (hunk 1 above)
Comment (jtpio) [reply setup.py]: :+1:

File: .github/workflows/main.yml (hunk 4 above)
Comment (maartenbreddels) [reply .github/workflows/main.yml]: Good idea, I'll do another pass of this PR.

**Comment by jtpio** [comment]
Opened https://github.com/jupyter/jupyter_server/pull/175 for the `extension_name` default and validation.

**Comment by jtpio** [comment]
I think we also need this change: https://github.com/jupyter/jupyter_server/pull/172

File: voila/app.py [hunk 5]
```diff
@@ -324,267 +269,120 @@ def _default_root_dir(self):
...
+                    recursive_update(conf['traitlet_configuration'], self.config.Voila)
+                    # pass merged config to overall voila config
+                    self.extension_config.Voila = Config(conf['traitlet_configuration'])
+        self.serverapp.log.debug("using template: %s", self.template)
+        self.serverapp.log.debug("nbconvert template paths:\n\t%s", "\n\t".join(self.nbconvert_template_paths))
+        self.serverapp.log.debug("template paths:\n\t%s", "\n\t".join(self.template_paths))
+        self.serverapp.log.debug("static paths:\n\t%s", "\n\t".join(self.static_paths))
 
         jenv_opt = {"autoescape": True}  # we might want extra options via cmd line like notebook server
         env = jinja2.Environment(loader=jinja2.FileSystemLoader(self.template_paths), extensions=['jinja2.ext.i18n'], **jenv_opt)
```
Comment (jtpio) [inline voila/app.py]: I wonder if we could make use of `ExtensionAppJinjaMixin` here: https://github.com/jupyter/jupyter_server/blob/5f99e53f7206c1272494a977e3246aa870d74fdf/jupyter_server/extension/application.py#L89

**Comment by davidbrochart** [comment]
I'm trying to sync this PR with master, I opened a PR in @maartenbreddels's  `refactor_extension_app` branch: https://github.com/maartenbreddels/voila/pull/1

**Comment by stonebig** [comment]
hello,

I'm experimenting Jupyter-server-1.0.1, and pip warns me voila is not compatible yet.
```
voila 0.2.2 requires jupyter-server<0.4.0,>=0.3.0, but you'll have jupyter-server 1.0.1 which is incompatible.
```

Is there an ongoing set of  pull request that resolves this ?

**Comment by trungleduc** [comment]
Closing as the proposed changes are not applicable anymore to the next release of Voila

**Pull request closed without merging**